File: python_backend/services/npi_service.py
from __future__ import annotations

import json
import re
from typing import Dict, Optional

import requests

CMS_NPI_REGISTRY_URL = "https://npiregistry.cms.hhs.gov/api/"
CMS_API_VERSION = "2.1"


class NpiError(Exception):
    """Base exception for NPI operations."""


class NpiInvalidError(NpiError):
    """Raised when the provided NPI number is malformed."""


class NpiNotFoundError(NpiError):
    """Raised when the CMS registry has no results for the given NPI."""


class NpiLookupError(NpiError):
    """Raised when the CMS registry cannot be reached or returns an unexpected error."""


def normalize_npi(value: Optional[str]) -> str:
    digits = re.sub(r"[^0-9]", "", str(value or ""))
    return digits[:10]


def _extract_primary_taxonomy(record: Dict) -> Optional[str]:
    taxonomies = record.get("taxonomies") or []
    if not isinstance(taxonomies, list) or not taxonomies:
        return None

    def describe(taxonomy: Dict) -> Optional[str]:
        for key in ("desc", "classification", "specialization"):
            value = taxonomy.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
        return None

    primary = next(
        (tax for tax in taxonomies if str(tax.get("primary", "")).lower() == "true"),
        taxonomies[0],
    )
    return describe(primary)


def _format_name(basic: Dict) -> Optional[str]:
    if not isinstance(basic, dict):
        return None

    direct_name = basic.get("name")
    if isinstance(direct_name, str) and direct_name.strip():
        return direct_name.strip()

    parts = [
        basic.get("first_name"),
        basic.get("middle_name"),
        basic.get("last_name"),
    ]
    parts = [part.strip() for part in parts if isinstance(part, str) and part.strip()]
    return " ".join(parts) if parts else None


def verify_npi(npi_number: str) -> Dict:
    normalized = normalize_npi(npi_number)
    if len(normalized) != 10:
        raise NpiInvalidError("NPI_INVALID")

    try:
        response = requests.get(
            CMS_NPI_REGISTRY_URL,
            params={"version": CMS_API_VERSION, "number": normalized},
            timeout=7,
        )
        response.raise_for_status()
        payload = response.json()
    except requests.RequestException as exc:
        raise NpiLookupError("NPI_LOOKUP_FAILED") from exc
    except ValueError as exc:
        raise NpiLookupError("NPI_LOOKUP_FAILED") from exc

    errors = payload.get("Errors")
    if errors:
        serialized = json.dumps(errors)
        if "No results" in serialized or "No Providers" in serialized:
            raise NpiNotFoundError("NPI_NOT_FOUND")
        if "Field number" in serialized or "number" in serialized.lower():
            raise NpiInvalidError("NPI_INVALID")
        raise NpiLookupError("NPI_LOOKUP_FAILED")

    results = payload.get("results") or []
    if not payload.get("result_count") or not results:
        raise NpiNotFoundError("NPI_NOT_FOUND")

    record = results[0]
    basic = record.get("basic") or {}

    return {
        "npiNumber": normalized,
        "name": _format_name(basic),
        "credential": basic.get("credential"),
        "enumerationType": basic.get("enumeration_type") or record.get("enumeration_type"),
        "organizationName": basic.get("organization_name"),
        "primaryTaxonomy": _extract_primary_taxonomy(record),
        "raw": record,
    }

File: backend_node_v1.9.21/server/services/npiService.js
const axios = require('axios');
const { logger } = require('../config/logger');

const CMS_NPI_REGISTRY_URL = 'https://npiregistry.cms.hhs.gov/api/';
const CMS_API_VERSION = '2.1';

const normalizeNpiNumber = (value) => {
  if (!value) {
    return '';
  }
  return String(value).replace(/[^0-9]/g, '').slice(0, 10);
};

const buildError = (message, status = 400, details) => {
  const error = new Error(message);
  error.status = status;
  if (details) {
    error.details = details;
  }
  return error;
};

const extractPrimaryTaxonomy = (taxonomies) => {
  if (!Array.isArray(taxonomies) || taxonomies.length === 0) {
    return null;
  }
  const primary =
    taxonomies.find((taxonomy) => String(taxonomy?.primary).toLowerCase() === 'true')
    || taxonomies[0];
  return (
    primary?.desc
    || primary?.classification
    || primary?.specialization
    || null
  );
};

const formatNameFromBasic = (basic) => {
  if (!basic) {
    return null;
  }
  if (basic.name && typeof basic.name === 'string') {
    return basic.name.trim() || null;
  }
  const parts = [basic.first_name, basic.middle_name, basic.last_name]
    .map((item) => (typeof item === 'string' ? item.trim() : ''))
    .filter(Boolean);
  return parts.length > 0 ? parts.join(' ') : null;
};

const verifyDoctorNpi = async (npiNumber) => {
  const normalized = normalizeNpiNumber(npiNumber);
  if (!/^\d{10}$/.test(normalized)) {
    throw buildError('NPI_INVALID', 400);
  }

  try {
    const response = await axios.get(CMS_NPI_REGISTRY_URL, {
      params: {
        version: CMS_API_VERSION,
        number: normalized,
      },
      timeout: 7000,
    });

    const { result_count: resultCount, results } = response?.data || {};
    if (!resultCount || !Array.isArray(results) || results.length === 0) {
      throw buildError('NPI_NOT_FOUND', 422);
    }

    const record = results[0];
    const basic = record?.basic || null;

    return {
      npiNumber: normalized,
      enumerationType: basic?.enumeration_type || record?.enumeration_type || null,
      name: formatNameFromBasic(basic),
      credential: basic?.credential || null,
      organizationName: basic?.organization_name || null,
      primaryTaxonomy: extractPrimaryTaxonomy(record?.taxonomies),
      raw: record,
    };
  } catch (error) {
    if (error?.message === 'NPI_INVALID' || error?.message === 'NPI_NOT_FOUND') {
      throw error;
    }

    const logPayload = {
      err: error,
      npiNumber: npiNumber ? String(npiNumber) : npiNumber,
    };

    if (axios.isAxiosError(error)) {
      logger.warn(logPayload, 'NPI registry lookup failed');
      throw buildError('NPI_LOOKUP_FAILED', 502, {
        status: error.response?.status,
        data: error.response?.data,
      });
    }

    logger.warn(logPayload, 'Unexpected error verifying NPI');
    throw buildError('NPI_LOOKUP_FAILED', 500);
  }
};

module.exports = {
  verifyDoctorNpi,
  normalizeNpiNumber,
};
